File: sdk/c-abi/examples/room_client.c
#include "fluvora.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void usage(void) {
    fputs(
        "usage: fluvora-c-demo <command> [arguments]\n"
        "  create <sfu|p2p|live|vod>\n"
        "  join <room-id>\n"
        "  chat <room-id> <text>\n"
        "  custom <room-id> <payload-json>\n"
        "  ice <room-id>\n"
        "  sfu-offer <room-id> <offer-sdp>\n"
        "  p2p-signal <room-id> <recipient-id> <kind> <payload-json>\n"
        "Environment: FLUVORA_BASE_URL and required FLUVORA_ACCESS_TOKEN\n",
        stderr);
}

static int print_result(int status, char *json) {
    if (status == FLUVORA_OK && json != NULL) {
        puts(json);
    } else {
        fprintf(stderr, "Fluvora operation failed with status %d\n", status);
    }
    fluvora_string_free(json);
    return status == FLUVORA_OK ? EXIT_SUCCESS : EXIT_FAILURE;
}

int main(int argc, char **argv) {
    const char *base_url = getenv("FLUVORA_BASE_URL");
    const char *token = getenv("FLUVORA_ACCESS_TOKEN");
    char *json = NULL;
    int status = FLUVORA_INVALID_ARGUMENT;
    int result = EXIT_FAILURE;
    int attempted = 0;

    if (argc < 2 || token == NULL || token[0] == '\0') {
        usage();
        return EXIT_FAILURE;
    }
    if (base_url == NULL || base_url[0] == '\0') {
        base_url = "http://127.0.0.1:8080";
    }

    FluvoraClient *client = fluvora_client_new(base_url, token);
    if (client == NULL) {
        fputs("failed to create Fluvora client\n", stderr);
        return EXIT_FAILURE;
    }

    if (strcmp(argv[1], "create") == 0 && argc == 3) {
        attempted = 1;
        status = fluvora_create_room(client, argv[2], 64, 16, &json);
    } else if (strcmp(argv[1], "join") == 0 && argc == 3) {
        attempted = 1;
        status = fluvora_join_room(client, argv[2], &json);
    } else if (strcmp(argv[1], "chat") == 0 && argc == 4) {
        attempted = 1;
        status = fluvora_send_chat(client, argv[2], "c-demo-message", argv[3], &json);
    } else if (strcmp(argv[1], "custom") == 0 && argc == 4) {
        attempted = 1;
        status = fluvora_send_custom_data(client, argv[2], "demo.c", 1, argv[3], &json);
    } else if (strcmp(argv[1], "ice") == 0 && argc == 3) {
        attempted = 1;
        status = fluvora_get_ice_configuration(client, argv[2], &json);
    } else if (strcmp(argv[1], "sfu-offer") == 0 && argc == 4) {
        /*
         * argv[3] is a complete ICE-gathered SDP string from the host PeerConnection.
         * Apply answer_sdp from the JSON result back to that same PeerConnection.
         */
        attempted = 1;
        status = fluvora_exchange_offer(client, argv[2], argv[3], &json);
    } else if (strcmp(argv[1], "p2p-signal") == 0 && argc == 6) {
        attempted = 1;
        status = fluvora_post_signal(client, argv[2], argv[3], argv[4], argv[5], &json);
    } else {
        usage();
    }
    if (attempted != 0) {
        result = print_result(status, json);
    }

    fluvora_client_free(client);
    return result;
}

File: sdk/rust/examples/room_client.rs
//! Buildable control, signaling, and dependency-neutral WebRTC integration example.
//!
//! Every invocation reads the short-lived token from `FLUVORA_ACCESS_TOKEN`; tokens are never
//! accepted as command-line arguments because process listings and shell history are not secret.

use std::env;
use std::error::Error;
use std::fs;
use std::io;
use std::sync::{Arc, Mutex, PoisonError};

use fluvora_sdk::{CallbackWebRtcPeer, Client, RoomMode};

fn usage() {
    eprintln!(
        "usage: room_client <command> [arguments]\n\
         commands:\n\
           create <sfu|p2p|live|vod>\n\
           join <room-id>\n\
           chat <room-id> <text>\n\
           custom <room-id> <json>\n\
           ice <room-id>\n\
           sfu-offer <room-id> <offer.sdp> <answer.sdp>\n\
           p2p-signal <room-id> <recipient-id|-> <kind> <payload-json>\n\
           poll <room-id> [after]\n\
           leave <room-id>\n\
         environment:\n\
           FLUVORA_BASE_URL       default: http://127.0.0.1:8080\n\
           FLUVORA_ACCESS_TOKEN   required short-lived bearer token"
    );
}

fn argument(args: &[String], index: usize, name: &str) -> Result<String, io::Error> {
    args.get(index)
        .cloned()
        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, format!("missing {name}")))
}

fn room_mode(value: &str) -> Result<RoomMode, io::Error> {
    match value {
        "sfu" => Ok(RoomMode::Sfu),
        "p2p" => Ok(RoomMode::P2p),
        "live" => Ok(RoomMode::Live),
        "vod" => Ok(RoomMode::Vod),
        _ => Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "mode must be sfu, p2p, live, or vod",
        )),
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        usage();
        return Err(io::Error::new(io::ErrorKind::InvalidInput, "missing command").into());
    }
    let base_url =
        env::var("FLUVORA_BASE_URL").unwrap_or_else(|_| "http://127.0.0.1:8080".to_owned());
    let token = env::var("FLUVORA_ACCESS_TOKEN")
        .map_err(|_| io::Error::new(io::ErrorKind::NotFound, "FLUVORA_ACCESS_TOKEN is required"))?;
    let client = Client::new(base_url, token)?;
    execute_command(&client, &args).await
}

async fn execute_command(client: &Client, args: &[String]) -> Result<(), Box<dyn Error>> {
    match args[1].as_str() {
        "create" => {
            let mode = room_mode(&argument(args, 2, "mode")?)?;
            let room = client.create_room(mode, Some(64), Some(16)).await?;
            println!("{}", serde_json::to_string_pretty(&room)?);
        }
        "join" => {
            let result = client.join(&argument(args, 2, "room-id")?).await?;
            println!("{}", serde_json::to_string_pretty(&result)?);
        }
        "chat" => {
            let room_id = argument(args, 2, "room-id")?;
            let text = argument(args, 3, "text")?;
            let message_id = format!("rust-demo-{}", std::process::id());
            let result = client.send_chat(&room_id, &message_id, &text).await?;
            println!("{}", serde_json::to_string_pretty(&result)?);
        }
        "custom" => {
            let room_id = argument(args, 2, "room-id")?;
            let payload = serde_json::from_str(&argument(args, 3, "json")?)?;
            let result = client
                .send_custom_data(&room_id, "demo.rust", 1, payload)
                .await?;
            println!("{}", serde_json::to_string_pretty(&result)?);
        }
        "ice" => {
            let result = client
                .get_ice_configuration(&argument(args, 2, "room-id")?)
                .await?;
            println!("{}", serde_json::to_string_pretty(&result)?);
        }
        "sfu-offer" => {
            connect_sfu_from_file(client, args).await?;
        }
        "p2p-signal" => {
            let room_id = argument(args, 2, "room-id")?;
            let recipient = argument(args, 3, "recipient-id")?;
            let kind = argument(args, 4, "kind")?;
            let payload = serde_json::from_str(&argument(args, 5, "payload-json")?)?;
            let signal = client
                .post_signal(
                    &room_id,
                    (recipient != "-").then_some(recipient),
                    kind,
                    payload,
                )
                .await?;
            println!("{}", serde_json::to_string_pretty(&signal)?);
        }
        "poll" => {
            let room_id = argument(args, 2, "room-id")?;
            let after = args.get(3).map_or(Ok(0), |value| value.parse::<u64>())?;
            let result = client.poll_signals(&room_id, after).await?;
            println!("{}", serde_json::to_string_pretty(&result)?);
        }
        "leave" => {
            let result = client.leave(&argument(args, 2, "room-id")?).await?;
            println!("{}", serde_json::to_string_pretty(&result)?);
        }
        _ => {
            usage();
            return Err(io::Error::new(io::ErrorKind::InvalidInput, "unknown command").into());
        }
    }
    Ok(())
}

async fn connect_sfu_from_file(client: &Client, args: &[String]) -> Result<(), Box<dyn Error>> {
    let room_id = argument(args, 2, "room-id")?;
    let offer_path = argument(args, 3, "offer.sdp")?;
    let answer_path = argument(args, 4, "answer.sdp")?;
    let offer = fs::read_to_string(offer_path)?;
    let answer_slot = Arc::new(Mutex::new(None::<String>));
    let answer_writer = Arc::clone(&answer_slot);
    let mut peer = CallbackWebRtcPeer::new(
        move || {
            let value = offer.clone();
            Box::pin(async move { Ok(value) })
        },
        move |answer| {
            let destination = Arc::clone(&answer_writer);
            Box::pin(async move {
                *destination.lock().unwrap_or_else(PoisonError::into_inner) = Some(answer);
                Ok(())
            })
        },
    )
    .with_room_data_channel(|| {
        // Create reliable/ordered `fluvora.room.v1` in the host WebRTC engine here.
        Box::pin(async { Ok(()) })
    });
    let session = client.connect_sfu(&room_id, &mut peer).await?;
    let answer = answer_slot
        .lock()
        .unwrap_or_else(PoisonError::into_inner)
        .take()
        .ok_or_else(|| io::Error::other("server answer callback was not invoked"))?;
    fs::write(answer_path, answer)?;
    println!("{}", serde_json::to_string_pretty(&session)?);
    Ok(())
}
